custom-login/app.js

```diff
@@ -0,0 +1,255 @@
+					$window.localStorage["image"] = res.data.image;
+					$window.localStorage["imageName"] = res.data.name;
+					$scope.img = res.data.image;
+					$scope.imgName = res.data.name;
+					$location.hash("imgAnchor");
+					$anchorScroll();
+				}
+			});
+		}
+
+		/**
+		 *	Refreshes the current session 
+		 */
+		$scope.refreshSession = function() {
+			oktaAuth.refreshSession()
+			.then(function(res){
+				$window.localStorage["sessionObject"] = angular.toJson(res);
+				refresh(100);
+				$location.hash("sessionAnchor");
+	  			$anchorScroll();
+			}, function(err){
+				console.log(err);
+			});
+		}
+		
+		/**
+		 *	Closes the current session
+		 */
+		$scope.closeSession = function() {
+			oktaAuth.closeSession()
+			.then(function(res){
+				$window.localStorage["session"] = false;
+				refresh(100);
+				$location.hash("userResponseAnchor");
+	  			$anchorScroll();
+	  		}, function(err){
+	  			console.log(err)
+	  		});
+		}
+
+		/**
+		 *	Refreshes Id token and updates $scope
+		 */
+		$scope.renewIdToken = function() {
+			// The access token cannot be refreshed - cached access token saved
+			oktaAuth.renewIdToken(clientScopes)
+			.then(function(res){
+				$window.localStorage["userInfo"] = angular.toJson({
+					"idToken" : res.idToken,
+					"accessToken" : JSON.parse($window.localStorage["userInfo"]).accessToken
+				});
+				refresh(100);
+		      	$location.hash("userInfoAnchor");
+	  		  	$anchorScroll();
+			}, function(err){
+				// handle error
+			});
+	  	}
+
+	  	/**
+	  	 *	Decodes the current Id token
+	  	 */
+	  	$scope.decodeIdToken = function(token) {
+	  		oktaAuth.decodeIdToken(token)
+	  		.then(function(res){
+	  			$window.localStorage["decodedIdToken"] = angular.toJson(res);
+	  			refresh(100);
+	  			$location.hash("decodeAnchor");
+	  			$anchorScroll();
+	  		}, function(err){
+	  			console.error(err + " can not be decoded");
+	  		});
+	  	}
+
+	  	/**
+		 *	Refreshes the current page given time duration until refresh
+	  	 */
+		function refresh(duration){
+			setTimeout(function() {$route.reload();}, duration);
+		}
+
+		/**
+		 *	Clears the localStorage saved in the web browser and scope variables
+		 */
+		function clearStorage(){
+			$window.localStorage.clear();
+			$scope = $scope.$new(true);
+		}
+
+		/**
+		 *	Signout method called via button selection
+		 */
+		$scope.signout = function() {
+			$timeout(function() {
+				oktaAuth.signout()
+				.then(function(res){
+					console.log(res);
+					clearStorage();
+					$location.url("/login");
+				}, function(err){
+					console.log(err);
+				});
+			}, 100);
+		}
+});
+
+/**
+ *	Authenticates the user with custom login UI using AuthSDK
+ *
+ * 	Stores the response object in localStorage and sets the current session to true
+ */
+app.controller("LoginController",
+	function($scope, $window, $location, $timeout, $route, authClient){
+		$scope.authenticate = function(user) {
+			var res = authClient.login(user.email, user.password);
```

Comment:
It's a bit confusing to call this return value "res" and then use res as the result of the promise in the next line